tap_freshdesk/client.py

```diff
@@ -7,7 +7,113 @@
 
 LOGGER = singer.get_logger()
 BASE_URL = "https://{}.freshdesk.com"
+REQUEST_TIMEOUT = 300
 
+class FreshdeskException(Exception):
+    pass
+
+class FreshdeskValidationError(FreshdeskException):
+    pass
+
+class FreshdeskAuthenticationError(FreshdeskException):
+    pass
+
+class FreshdeskAccessDeniedError(FreshdeskException):
+    pass
+
+class FreshdeskNotFoundError(FreshdeskException):
+    pass
+
+class FreshdeskMethodNotAllowedError(FreshdeskException):
+    pass
+
+class FreshdeskUnsupportedAcceptHeaderError(FreshdeskException):
+    pass
+
+class FreshdeskConflictingStateError(FreshdeskException):
+    pass
+
+class FreshdeskUnsupportedContentError(FreshdeskException):
+    pass
+
+class FreshdeskRateLimitError(FreshdeskException):
+    pass
+
+class Server5xxError(FreshdeskException):
+    pass
+
+class FreshdeskServerError(Server5xxError):
+    pass
+
+
+ERROR_CODE_EXCEPTION_MAPPING = {
+    400: {
+        "raise_exception": FreshdeskValidationError,
+        "message": "The request body/query string is not in the correct format."
+    },
+    401: {
+        "raise_exception": FreshdeskAuthenticationError,
+        "message": "The Authorization header is either missing or incorrect."
+    },
+    403: {
+        "raise_exception": FreshdeskAccessDeniedError,
+        "message": "The agent whose credentials were used to make this request was not authorized to perform this API call."
+    },
+    404: {
+        "raise_exception": FreshdeskNotFoundError,
+        "message": "The request contains invalid ID/Freshdesk domain in the URL or an invalid URL itself."
+    },
+    405: {
+        "raise_exception": FreshdeskMethodNotAllowedError,
+        "message": "This API request used the wrong HTTP verb/method."
+    },
+    406: {
+        "raise_exception": FreshdeskUnsupportedAcceptHeaderError,
+        "message": "Only application/json and */* are supported in 'Accepted' header."
+    },
+    409: {
+        "raise_exception": FreshdeskConflictingStateError,
+        "message": "The resource that is being created/updated is in an inconsistent or conflicting state."
+    },
+    415: {
+        "raise_exception": FreshdeskUnsupportedContentError,
+        "message": "Content type application/xml is not supported. Only application/json is supported."
+    },
+    429: {
+        "raise_exception": FreshdeskRateLimitError,
+        "message": "The API rate limit allotted for your Freshdesk domain has been exhausted."
+    },
+    500: {
+        "raise_exception": FreshdeskServerError,
+        "message": "Unexpected Server Error."
+    },
+}
+
+def raise_for_error(response):
+    """
+    Retrieve the error code and the error message from the response and return custom exceptions accordingly.
+    """
+    try:
```

Comment:
Why do we need this? We can call this function if status_code is not equal to 200

Reply:
Agree @dbshah1212. Removed `try-except` and raising error directly if `status_code` is not equal to 200.